openedx/features/enterprise_support/api.py

```diff
@@ -311,34 +336,51 @@ def enterprise_customer_for_request(request):
             # Check if there's an Enterprise Customer such that the linked SSO provider
             # has an ID equal to the ID we got from the running pipeline or from the
             # request tpa_hint URL parameter.
-            ec_uuid = EnterpriseCustomer.objects.get(
+            enterprise_customer_uuid = EnterpriseCustomer.objects.get(
                 enterprise_customer_identity_provider__provider_id=sso_provider_id
             ).uuid
         except EnterpriseCustomer.DoesNotExist:
             # If there is not an EnterpriseCustomer linked to this SSO provider, set
             # the UUID variable to be null.
-            ec_uuid = None
+            enterprise_customer_uuid = None
     else:
         # Check if we got an Enterprise UUID passed directly as either a query
         # parameter, or as a value in the Enterprise cookie.
-        ec_uuid = request.GET.get('enterprise_customer') or request.COOKIES.get(settings.ENTERPRISE_CUSTOMER_COOKIE_NAME)
+        enterprise_customer_uuid = request.GET.get('enterprise_customer') or request.COOKIES.get(
+            settings.ENTERPRISE_CUSTOMER_COOKIE_NAME
+        )
 
-    if not ec_uuid and request.user.is_authenticated():
+    if not enterprise_customer_uuid and request.user.is_authenticated():
         # If there's no way to get an Enterprise UUID for the request, check to see
         # if there's already an Enterprise attached to the requesting user on the backend.
         learner_data = get_enterprise_learner_data(request.site, request.user)
         if learner_data:
-            ec_uuid = learner_data[0]['enterprise_customer']['uuid']
-    if ec_uuid:
+            enterprise_customer_uuid = learner_data[0]['enterprise_customer']['uuid']
+    if enterprise_customer_uuid:
         # If we were able to obtain an EnterpriseCustomer UUID, go ahead
         # and use it to attempt to retrieve EnterpriseCustomer details
         # from the EnterpriseCustomer API.
+        enterprise_api_client = EnterpriseApiServiceClient()
+        enterprise_api_client_username = settings.ENTERPRISE_SERVICE_WORKER_USERNAME
+        if request.user.is_authenticated():
+            enterprise_api_client = EnterpriseApiClient(user=request.user)
+            enterprise_api_client_username = request.user.username
+
+        cache_key = get_cache_key(
+            site_domain=request.site.domain,
+            resource='get_enterprise_customer',
+            username=enterprise_api_client_username,
+        )
+        enterprise_customer = cache.get(cache_key)
+        if enterprise_customer:
+            return enterprise_customer
         try:
-            ec = EnterpriseApiClient().get_enterprise_customer(ec_uuid)
+            enterprise_customer = enterprise_api_client.get_enterprise_customer(enterprise_customer_uuid)
+            cache.set(cache_key, enterprise_customer, settings.ENTERPRISE_API_CACHE_TIMEOUT)
```

Comment:
We should only cache this if we made the API request with the service user. Now that I'm thinking about this more, I agree with @haikuginger's concerns involving a stale cache when we make this API request with the request user.